lib/logging.dart

```diff
@@ -364,3 +366,187 @@ class LogRecord {
+class LogManager {
+  static bool _initialized;
+
+  final Set<String> _enabledLoggers = Set<String>();
+
+  final StreamController<LogRecord> _logController =
+      StreamController.broadcast(sync: true);
+  final StreamController<String> _loggerAddedBroadcaster =
+      StreamController.broadcast();
+  final StreamController<String> _loggerEnabledBroadcaster =
+      StreamController.broadcast();
+
+  LogManager() {
+    Logger.root.onRecord.listen((record) {
+      if (shouldLog(record.loggerName)) {
+        _logController.add(record);
+      }
+    });
+  }
+
+  /// Returns a stream of messages added to loggers enabled by this manager.
+  Stream<LogRecord> get onRecord => _logController.stream;
+
+  /// Enable (or disable) logging of messages sent to the given [logger].
+  void enableLogging(String logger, {bool enable = true}) {
+    enable ? _enabledLoggers.add(logger) : _enabledLoggers.remove(logger);
+    _loggerEnabledBroadcaster.add(logger);
+  }
+
+  /// Whether messages to the given [logger] should be logged by this manager.
+  /// @see [enableLogging]
+  bool shouldLog(String logger) => _enabledLoggers.contains(logger);
+
+  /// Called to register service extensions.
+  void initServiceExtensions() {
+    // Avoid double initialization.
+    if (_initialized == true) {
+      return;
+    }
+
+    // Fire events for new loggers.
+    _loggerAddedBroadcaster.stream.listen((String name) {
+      developer.postEvent('logging.logger.added', <String, dynamic>{
+        'logger': name,
+      });
+    });
+
+    // Fire events for logger enablement changes.
+    _loggerEnabledBroadcaster.stream.listen((String name) {
+      developer.postEvent('logging.logger.enabled', <String, dynamic>{
+        'logger': name,
+        'enabled': shouldLog(name),
+      });
+    });
+
+    // Service for enabling loggers.
+    _registerServiceExtension(
+      name: 'enable',
+      callback: (Map<String, Object> parameters) async {
+        final String logger = parameters['logger'];
+        if (logger != null) {
+          if (parameters.containsKey('enabled')) {
+            enableLogging(logger, enable: parameters['enabled'] == 'true');
+          }
+          return <String, dynamic>{
+            'enabled': shouldLog(logger).toString(),
+          };
+        } else {
+          return <String, dynamic>{};
+        }
+      },
+    );
+
+    // Service for querying loggers.
+    _registerServiceExtension(
+      name: 'loggers',
+      callback: (Map<String, dynamic> parameters) async => {
+            'value': Logger._loggers.keys
+                .map((logger) => MapEntry(logger, <String, String>{
+                      'enabled': shouldLog(logger).toString(),
+                    }))
+          },
+    );
+
+    _initialized = true;
+  }
+
+  /// Registers a service extension method with the given name and a callback to
+  /// be called when the extension method is called.
+  void _registerServiceExtension({
```

Comment:
probably make this method public but protected so you can override it for testing.